Editorial: Reinstate async-evaluating state, with AsyncEvaluation as counter and initial state

guybedford opened the pull request:
This brings back an `async-evaluating` state which is true for any module either currently asynchronously executing, or waiting on a dependency that is currently asynchronously executing. The previous issue with this state was that we still needed a separate field for `AsyncEvaluating` due to handling cycle transitions, as discussed in https://github.com/tc39/proposal-top-level-await/pull/114.

We can bring this state back by adopting the state as well as still using the separate field for `AsyncEvaluating` which is renamed to `AsyncEvaluation`. Instead of treating execution completion as the transition of `AsyncEvaluating` to false as before, we then treat execution completion as the transition of the `async-evaluating` state to `evaluated`.

This PR can be seen as performing the following (non-normative) replacement operations in order:
* Replace all `~evaluated~` state checks with a check for `~evaluated~` or `~evaluating-async~`.
* Replace the `~evaluated~` transition for async executions with an `~evaluating-async~` transition.
* Rename `[[AsyncEvaluating]]` to `[[AsyncEvaluation]]` with an initial `~empty~` state.
* Include a transition of `[[State]]` from `evaluating-async -> evaluating` whenever we previously set `[[AsyncEvaluating]]` to false.
* Never transition `[[AsyncEvaluation]]` back to false, and instead entirely use the `[[State]]` transitions for these checks.

The `AsyncEvaluation` field provides three purposes:

1. It indicates which modules have async dependencies.
2. It acts as the counter for async execution ordering.
3. Most importantly, during the initial execution traversal, it allows cycles to attach to the completion of other modules in the cycle by knowing that they are currently executing. Because cycles transition their states together from `~evaluating~` to `~async-evaluating~`, we wouldn't otherwise have this information without some other indicator of a module of the same cycle that we need to attach to for async completion (we know it's evaluating because the "evaluating" state is set early in traversal, but the question is if its evaluating and its execution has already been triggered, which otherwise doesn't have a state representation). This was the underlying cause of previous bugs like https://github.com/tc39/proposal-top-level-await/pull/109.

codehag commented:
Overall looks reasonable to me. A couple of comments, as i can't comment directly:

This is a nit, but might be important later on: `AsyncEvaluation` is a bit ambiguous. The module isn't strictly async itself. What it indicates is that this module has an async job that it depends on, either in its own top level code or in its dependencies. Maybe something like `AwaitsAsyncJobs` or something to this effect (`QueueEvaluation` maybe? its a bit distant from async but it also gets the idea across) -- something that makes it clear that this isn't related to this module itself being async.

iirc implementations transition back to a false state. I think this was important but can't remember why right now

We need to also update the tables. AsyncEvaluation (or whatever we call it, shouldn't be set to false on lines 1039, 1007, and 967.

As these are non-normative... I hope this won't require us pushing back one more meeting. I think what we should do is make two merging prs, one with the current spec text, one with this change included. The committee can decide if it wants to incorporate both in stage 4, or split this out as a clarification for later.

guybedford commented:
> AsyncEvaluation is a bit ambiguous.

Good point, I've renamed this to `QueuedEvaluation` ~~and also made another change to the meaning, to have it only apply to those module executions that are actually queued so the check becomes `_module_.[[Async]] or _module_.[[QueuedEvaluation]]` while keeping the simpler meanings of the fields~~. That all works out fine in the text now as well, let me know if that seems simpler? Seems like it was another common comment here...

> iirc implementations transition back to a false state. I think this was important but can't remember why right now

This is handled by replacing all checks for when AsyncEvaluating is false again with the `~evaluated~` state itself. Mainly this is needed for completion checks, so evaluated as the primary completion check is one of the refactoring benefits here (down to still needing the cycle root check condition of course).

I've gone ahead and updated the tables.

codehag commented:
should [[State]] be [[Status]] in the example prose?

guybedford commented:
Actually separating `QueuedEvaluation` from async evaluation does cause the old cycle spec issue we had in the past, so I will keep the meaning of `QueuedEvaluation` as **any** async evaluation. Updating now with the other correction as well.
